osu.Game/Scoring/ScoreManager.cs

```diff
@@ -134,44 +134,46 @@ public async Task<long> GetTotalScoreAsync([NotNull] ScoreInfo score, ScoringMod
             if (string.IsNullOrEmpty(score.BeatmapInfo.MD5Hash))
                 return score.TotalScore;
 
-            int beatmapMaxCombo;
+            int? beatmapMaxCombo = await GetBeatmapMaximumComboAsync(score, cancellationToken).ConfigureAwait(false);
+            if (beatmapMaxCombo == null)
+                return score.TotalScore;
+
+            if (beatmapMaxCombo == 0)
+                return 0;
 
+            var ruleset = score.Ruleset.CreateInstance();
+            var scoreProcessor = ruleset.CreateScoreProcessor();
+            scoreProcessor.Mods.Value = score.Mods;
+
+            return (long)Math.Round(scoreProcessor.ComputeFinalLegacyScore(mode, score, beatmapMaxCombo.Value));
+        }
+
+        /// <summary>
+        /// Retrieves the maximum achievable combo on the beatmap of the provided score.
+        /// </summary>
+        /// <param name="score">The <see cref="ScoreInfo"/> to compute the maximum achievable combo for.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> to cancel the process.</param>
+        /// <returns>The maximum achievable combo.</returns>
+        public async Task<int?> GetBeatmapMaximumComboAsync([NotNull] ScoreInfo score, CancellationToken cancellationToken = default)
```

Comment:
Method has weird name - talks about "beatmap maximum combo" but accepts score. Should be maximum combo for score, I'd think.

Reply:
Would conflict with `score.MaxCombo` which is the main reason why it's not called that. Alternative would be `GetMaximumAchievableComboAsync` (initial name I had), which makes sense since it depends on whether the score is legacy/non-legacy, not just accessing `score.BeatmapInfo` directly.

Reply:
"maximum achievable" is better yeah.